src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs

```diff
@@ -262,28 +264,89 @@ public string Name
                 {
                     if (IsConstructedGenericType)
                     {
-                        _name = TypeNameParserHelpers.GetName(GetGenericTypeDefinition().FullName.AsSpan()).ToString();
+                        _name = GetGenericTypeDefinition().Name;
                     }
                     else if (IsPointer || IsByRef || IsArray)
                     {
                         ValueStringBuilder builder = new(stackalloc char[64]);
                         builder.Append(GetElementType().Name);
                         _name = TypeNameParserHelpers.GetRankOrModifierStringRepresentation(_rankOrModifier, ref builder);
                     }
-                    else if (_nestedNameLength > 0 && _fullName is not null)
-                    {
-                        _name = TypeNameParserHelpers.GetName(_fullName.AsSpan(0, _nestedNameLength)).ToString();
-                    }
                     else
                     {
-                        _name = TypeNameParserHelpers.GetName(FullName.AsSpan()).ToString();
+                        // _fullName can be null only in constructed generic or modified types, which we handled above.
+                        Debug.Assert(_fullName is not null);
+                        ReadOnlySpan<char> name = _fullName.AsSpan();
+                        if (_nestedNameLength > 0)
+                        {
+                            name = name.Slice(0, _nestedNameLength);
+                        }
+                        if (IsNested)
+                        {
+                            // If the type is nested, we know the length of the declaring type's full name.
+                            // Get the characters after that plus one for the '+' separator.
+                            name = name.Slice(_declaringType._nestedNameLength + 1);
+                        }
+                        else if (TypeNameParserHelpers.IndexOfNamespaceDelimiter(name) is int idx && idx >= 0)
+                        {
+                            // If the type is not nested, find the namespace delimiter in the full name and and return the substring after it.
+                            name = name.Slice(idx + 1);
+                        }
+                        _name = name.ToString();
                     }
                 }
 
                 return _name;
             }
         }
 
+        /// <summary>
+        /// The namespace of this type; e.g., "System".
+        /// </summary>
+        /// <exception cref="InvalidOperationException">This instance is a nested type.</exception>
+        public string Namespace
+        {
+            get
+            {
+                if (IsNested)
```

Comment:
Should this check be done on rootTypeName?

It does not make sense for "MyNamespace.A+B[]" to return `MyNamespace` when "MyNamespace.A+B" fails with InvalidOperationException.

Reply:
Would it make sense to throw in arrays, pointers and refs, like in nested types?

Reply:
I do not have a strong opinion. 

If I had to choose, I would throw for nested types only. I think that the pit of failure that we want to protect from is limited to nested types. If you have `MyNamespace.A[]`, it makes sense that `Name` returns `A[]` and `Namespace` returns `MyNamespace`.

Reply:
Makes sense, done.